File: code/testScript/benchmark.py
import os
import subprocess
import time
import resource

runtTestsXTimes = 5


def main():
    print("Starting Benchmarking Script\n\r")
    clean = ["Clean", 'java']

    # ##### Dynamic Taint Tracking Tools #####
    # Dynamic Taint Tracker
    dtpXboot = "-Xbootclasspath/p:/home/fredrik/Documents/Omegapoint/masterthesis-fredrik/code/dtp/build/libs/dtp-rt-1.0-SNAPSHOT.jar"
    dtpAgent = "-javaagent:/home/fredrik/Documents/Omegapoint/masterthesis-fredrik/code/dtp/build/libs/dtp-agent-1.0-SNAPSHOT.jar"
    dynamicTaintTracker = ["Dynamic Taint Tracker", 'java', dtpXboot, dtpAgent]
    # -Xbootclasspath/p:/home/fredrik/Documents/Omegapoint/masterthesis-fredrik/code/dtp/build/libs/dtp-rt-1.0-SNAPSHOT.jar -javaagent:/home/fredrik/Documents/Omegapoint/masterthesis-fredrik/code/dtp/build/libs/dtp-agent-1.0-SNAPSHOT.jar

    # Dynamic Security Taint Propagation DONT WORK!!!
    stpXboot = "-Xbootclasspath/p:/home/fredrik/Documents/Omegapoint/Benchmarking/DynamicTaintTrackers/security_taint_propagation/security_taint_extension/target/tainted-rt-1.8.jar"
    stpAgent = "-javaagent:/home/fredrik/Documents/Omegapoint/Benchmarking/DynamicTaintTrackers/security_taint_propagation/aspectjweaver-1.9.1.jar"
    securityTaintPropagation = [
        "Dynamic Security Taint Propagation", 'java',   stpXboot, stpAgent]
    # -Xbootclasspath/p:/home/fredrik/Documents/Omegapoint/Benchmarking/Dynamic\ Taint\ Trackers/security_taint_propagation/security_taint_extension/target/tainted-rt-1.8.jar -javaagent:/home/fredrik/Documents/Omegapoint/Benchmarking/Dynamic\ Taint\ Trackers/security_taint_propagation/aspectjweaver-1.9.1.jar

    # Phosphor
    phosJava = "/home/fredrik/Documents/Omegapoint/Benchmarking/DynamicTaintTrackers/phosphor-master/Phosphor/target/jre-inst/bin/java"
    phosXboot = "-Xbootclasspath/a:/home/fredrik/Documents/Omegapoint/Benchmarking/DynamicTaintTrackers/phosphor-master/Phosphor/target/Phosphor-0.0.4-SNAPSHOT.jar"
    phosAgent = "-javaagent:/home/fredrik/Documents/Omegapoint/Benchmarking/DynamicTaintTrackers/phosphor-master/Phosphor/target/Phosphor-0.0.4-SNAPSHOT.jar"
    phosphor = ["Phosphor", phosJava, phosXboot, phosAgent]

    # ##### Test Suits #####
    # DeCapo
    daCapo = [
        '-jar', "/home/fredrik/Documents/Omegapoint/Benchmarking/TestSuites/dacapo-9.12-MR1-bach.jar"]
    # daCapoPrograms = ["avrora"]
    daCapoPrograms = ["avrora", "eclipse", "fop",  "h2", "jython", "luindex",
                      "lusearch", "pmd", "sunflow", "tomcat", "tradebeans", "tradesoap", "xalan"]

    print("##### ADDED OVERHEAD #####")
    for daCapoProgram in daCapoPrograms:
        print("\n\r\t# DaCapo " + daCapoProgram)

        # for trackingTools in [clean]:
        for trackingTools in [clean, dynamicTaintTracker]:
            [text, *rest] = trackingTools

            daCapoExec = [text] + \
                rest + daCapo + [daCapoProgram]
            print(daCapoExec)
            measureTime(*daCapoExec)

    print("\n\rEnd Benchmarking Script")


def current_milli_time(): return int(round(time.time() * 1000))


def measureTime(text, *params):
    FNULL = open(os.devnull, 'w')

    averageTime = 0
    fastestTime = 0
    slowestTime = 0
    averageMem = 0
    fastestMem = 0
    slowestMem = 0
    for x in range(1, runtTestsXTimes + 1):
        print("", end="\r")
        print("\t\tLoading (", x, "/", runtTestsXTimes, ")", end='', flush=True)

        start_time = int(round(time.time() * 1000))

        p = subprocess.Popen(params, stdout=FNULL, stderr=subprocess.STDOUT)
        p.wait()
        time_diff = current_milli_time() - start_time

        retcode = p.returncode

        memUsageKilobytes = resource.getrusage(
            resource.RUSAGE_CHILDREN).ru_maxrss

        averageTime += time_diff / runtTestsXTimes
        averageMem += memUsageKilobytes / runtTestsXTimes

        if fastestTime < time_diff:
            fastestTime = time_diff
        if slowestTime > time_diff or slowestTime == 0:
            slowestTime = time_diff

        if fastestMem < memUsageKilobytes:
            fastestMem = memUsageKilobytes
        if slowestMem > memUsageKilobytes or slowestMem == 0:
            slowestMem = memUsageKilobytes

    print("", end="\r")
    print("\t\t" + text, "OK" if retcode == 0 else "ERROR",
          "                                                   ")
    if retcode == 0:
        print("\t\t\t Average", round(averageTime),
              "(ms) ", round(averageMem), "(kilobytes)")
        print("\t\t\t Fastest", round(fastestTime),
              "(ms) ", round(fastestMem), "(kilobytes)")
        print("\t\t\t Slowest", round(slowestTime),
              "(ms) ", round(slowestMem), "(kilobytes)")


if __name__ == "__main__":
    main()

File: code/testScript/benchmark.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <fcntl.h>
#include <time.h>

int main(int argc, char *argv[])
{
  pid_t childpid;
  struct rusage usage;
  struct timespec start, end;
  long mem_usage = 0, mem_average = 0, mem_min = 0, mem_max = 0;
  u_int64_t time_usage = 0, time_average = 0, time_min = 0,time_max = 0;
  int j, k, i, fd[2];
  int val = 0;

  int runtTestsXTimes = 10;
  const char *file = "java";

  // ##### Dynamic Taint Tracking Tools #####
  // # Dynamic Taint Tracker
  const char *dtpXboot = "-Xbootclasspath/p:/home/fredrik/Documents/Omegapoint/masterthesis-fredrik/code/dtp/build/libs/dtp-rt-1.0-SNAPSHOT.jar";
  const char *dtpAgent = "-javaagent:/home/fredrik/Documents/Omegapoint/masterthesis-fredrik/code/dtp/build/libs/dtp-agent-1.0-SNAPSHOT.jar";

  // ##### Test Suits #####
  // DeCapo
  const char *daCapo = "/home/fredrik/Documents/Omegapoint/Benchmarking/TestSuites/dacapo-9.12-MR1-bach.jar";
  const char *daCapoPrograms[] = {"avrora", "batik", "eclipse", "fop",  "h2", "jython", "luindex", "lusearch", "pmd", "sunflow", "tomcat", "tradebeans", "tradesoap", "xalan"};

  for (j = 0; j < sizeof(daCapoPrograms) / sizeof(daCapoPrograms[0]); j++) {
    printf("# DaCapo %s\n", daCapoPrograms[j]);
    printf("\tClean\n");
    
    for (i = 0 ; i < runtTestsXTimes; i++) {
      printf("\t\tLoading (%d / %d)", i + 1, runtTestsXTimes);
      fflush( stdout );
      // create pipe descriptors
      pipe(fd);

      childpid = fork();
      if(childpid != 0)  // parent
      {
        close(fd[1]);
        // read the data (blocking operation)
        read(fd[0], &mem_usage, sizeof(mem_usage));
        read(fd[0], &time_usage, sizeof(time_usage));

        mem_average += mem_usage / runtTestsXTimes;
        time_average += time_usage / runtTestsXTimes;

        if(mem_usage > mem_max) mem_max = mem_usage;
        if(mem_usage < mem_min || mem_min == 0) mem_min = mem_usage;

        if(time_usage > time_max) time_max = time_usage;
        if(time_usage < time_min || time_min == 0) time_min = time_usage;

        // close the read-descriptor
        close(fd[0]);
      }
      else  // child
      {    
        close(fd[0]); // writing only, no need for read-descriptor
        
        int returnStatus;  
        
        clock_gettime(CLOCK_MONOTONIC_RAW, &start);
        childpid = fork();
        if(childpid != 0)  // parent
        {
          waitpid(childpid, &returnStatus, 0);
          clock_gettime(CLOCK_MONOTONIC_RAW, &end);
          time_usage = (end.tv_sec - start.tv_sec) * 1000000 + (end.tv_nsec - start.tv_nsec) / 1000;
        }
        else  // child
        {
          int fd2 = open("/dev/null", O_WRONLY);
          dup2(fd2, 1);
          dup2(fd2, 2);
          close(fd2);
          execl("/usr/bin/java", "java", "-jar", daCapo, daCapoPrograms[j], NULL);
        }

        getrusage(RUSAGE_CHILDREN, &usage);
        mem_usage = usage.ru_maxrss;

        write(fd[1], &mem_usage, sizeof(mem_usage)); // send the value on the write-descriptor
        write(fd[1], &time_usage, sizeof(time_usage)); // send the value on the write-descriptor

        close(fd[1]); // close the write descriptor
        return mem_usage;
      }
    }

    printf("\r\t\tTime (microseconds)\n");
    printf("\t\t\tAverage: %ld\n", time_average);
    printf("\t\t\tMin: %ld\n", time_min);
    printf("\t\t\tMax: %ld\n", time_max);

    printf("\t\tMemory (kilobytes)\n");
    printf("\t\t\tAverage: %ld\n", mem_average);
    printf("\t\t\tMin: %ld\n", mem_min);
    printf("\t\t\tMax: %ld\n", mem_max);

    time_average = time_min = time_max = 0;
    mem_average = mem_min = mem_max = 0;

    printf("\tDynamic Taint Tracking\n");
    
    for (i = 0 ; i < runtTestsXTimes; i++) {
      printf("\t\tLoading (%d / %d)", i + 1, runtTestsXTimes);
      fflush( stdout );
      // create pipe descriptors
      pipe(fd);

      childpid = fork();
      if(childpid != 0)  // parent
      {
        close(fd[1]);
        // read the data (blocking operation)
        read(fd[0], &mem_usage, sizeof(mem_usage));
        read(fd[0], &time_usage, sizeof(time_usage));

        mem_average += mem_usage / runtTestsXTimes;
        time_average += time_usage / runtTestsXTimes;

        if(mem_usage > mem_max) mem_max = mem_usage;
        if(mem_usage < mem_min || mem_min == 0) mem_min = mem_usage;

        if(time_usage > time_max) time_max = time_usage;
        if(time_usage < time_min || time_min == 0) time_min = time_usage;

        // close the read-descriptor
        close(fd[0]);
      }
      else  // child
      {    
        close(fd[0]); // writing only, no need for read-descriptor
        
        int returnStatus;  
        
        clock_gettime(CLOCK_MONOTONIC_RAW, &start);
        childpid = fork();
        if(childpid != 0)  // parent
        {
          waitpid(childpid, &returnStatus, 0);
          clock_gettime(CLOCK_MONOTONIC_RAW, &end);
          time_usage = (end.tv_sec - start.tv_sec) * 1000000 + (end.tv_nsec - start.tv_nsec) / 1000;
        }
        else  // child
        {
          int fd2 = open("/dev/null", O_WRONLY);
          dup2(fd2, 1);
          dup2(fd2, 2);
          close(fd2);
          execl("/usr/bin/java", "java", dtpXboot, dtpAgent, "-jar", daCapo, daCapoPrograms[j], NULL);
        }

        getrusage(RUSAGE_CHILDREN, &usage);
        mem_usage = usage.ru_maxrss;

        write(fd[1], &mem_usage, sizeof(mem_usage)); // send the value on the write-descriptor
        write(fd[1], &time_usage, sizeof(time_usage)); // send the value on the write-descriptor

        close(fd[1]); // close the write descriptor
        return mem_usage;
      }
    }

    printf("\r\t\tTime (microseconds)\n");
    printf("\t\t\tAverage: %ld\n", time_average);
    printf("\t\t\tMin: %ld\n", time_min);
    printf("\t\t\tMax: %ld\n", time_max);

    printf("\t\tMemory (kilobytes)\n");
    printf("\t\t\tAverage: %ld\n", mem_average);
    printf("\t\t\tMin: %ld\n", mem_min);
    printf("\t\t\tMax: %ld\n\n", mem_max);

    time_average = time_min = time_max = 0;
    mem_average = mem_min = mem_max = 0;
  }

}
